lib/puma/server.rb

```diff
@@ -320,41 +320,65 @@ def handle_servers
         check = @check
         sockets = [check] + @binder.ios
         pool = @thread_pool
+        binder = @binder
         queue_requests = @queue_requests
         drain = @options[:drain_on_shutdown] ? 0 : nil
 
-        addr_send_name, addr_value = case @options[:remote_address]
-        when :value
-          [:peerip=, @options[:remote_address_value]]
-        when :header
-          [:remote_addr_header=, @options[:remote_address_header]]
-        when :proxy_protocol
-          [:expect_proxy_proto=, @options[:remote_address_proxy_protocol]]
-        else
-          [nil, nil]
-        end
-
-        while @status == :run || (drain && shutting_down?)
+        # option to local variables
+        # Using local variables avoid calling `@options#[]` method each time we need the value
+        remote_address_option = @options[:remote_address]
+        remote_address_value_option = @options[:remote_address_value]
+        remote_address_header_option = @options[:remote_address_header]
+        remote_address_proxy_protocol_option = @options[:remote_address_proxy_protocol]
+        wait_for_less_busy_worker = @options[:wait_for_less_busy_worker]
+        wait_for_less_busy_worker_enabled = !@options[:wait_for_less_busy_worker].nil?
+
+        is_shutting_down = false
+        while @status == :run || (drain && (is_shutting_down = shutting_down?))
           begin
-            ios = IO.select sockets, nil, nil, (shutting_down? ? 0 : nil)
+            ios = IO.select(sockets, nil, nil, (is_shutting_down ? 0 : nil))
             break unless ios
             ios.first.each do |sock|
               if sock == check
                 break if handle_check
               else
                 pool.wait_until_not_full
-                pool.wait_for_less_busy_worker(@options[:wait_for_less_busy_worker])
 
-                io = begin
-                  sock.accept_nonblock
-                rescue IO::WaitReadable
-                  next
+                # Wait for less busy worker?
+                if wait_for_less_busy_worker_enabled
```

Comment:
Maybe remove `if wait_for_less_busy_worker_enabled` and just use:
```ruby
if wait_for_less_busy_worker
```
It should either be a positive number or nil?  Both are checked in the method...